File: TopCoder/SRM/636/ChocolateDividingEasy/Solution.java
//-----------------------------------------------------------------------
// <copyright file="Solution.cs" company="MyCompany">
//     Copyright (c) MyCompany. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
package topcoder.srm636.ChocolateDividingEasy;
import java.util.*;

public class Solution {
	public static void main(String[] args)
	{
		int[][] grid = new int[3][4];
		grid[0] = new int[]{9, 7, 6, 8};
		grid[1] = new int[]{6, 7, 6, 7};
		grid[2] = new int[]{5, 3, 1, 3};
		
		int numVerticalCuts = 2;
		int numHorizontalCuts = 2;
		solve(grid, numVerticalCuts, numHorizontalCuts);
		System.out.println("Ans: " + bestMinPiece);
	}

	private static int bestMinPiece = Integer.MIN_VALUE;
	private static int solve(
			int[][] grid, int numVerticalCuts, int numHorizontalCuts) {		
		
		List<List<Integer>> vCutsList = getCuts(grid.length, numVerticalCuts);
		List<List<Integer>> hCutsList = getCuts(grid[0].length, numHorizontalCuts);
		
		for(List<Integer> vCuts:vCutsList){
			for(List<Integer> hCuts:hCutsList){
				calculateBestMinPiece(grid, hCuts, vCuts);
			}
		}
					
		return bestMinPiece;
	}
	private static List<List<Integer>> getCuts(int length, int numCuts) {
		List<Integer> set = new ArrayList<Integer>();
		for(int j=1; j<length; j++)
		{
			set.add(j);
		}
		List<List<Integer>> results = 
				new ArrayList<List<Integer>>();
		int n = (int)Math.pow(2, length-1);		
		for(int i=0; i<n; i++)
		{
			int k=i;
			int index = 0;
			List<Integer> combo = new ArrayList<Integer>();
			while((k&1)!=0){	
				combo.add(set.get(index));
				k>>=1;
		        ++index;
			}
			if(combo.size()==numCuts){
				results.add(combo);
			}
		}
		return results;
	}
	private static void calculateBestMinPiece(int[][] grid,
			List<Integer> hCuts1, List<Integer> vCuts1) {

		System.out.print("Vertical cuts: ");
		for(int v:vCuts1){
			System.out.print(v + " ");
		}
		System.out.println();
		
		System.out.print("Horizontal cuts: ");
		for(int h:hCuts1){
			System.out.print(h + " ");
		}
		System.out.println();
		System.out.println();
						
		ArrayList<Integer> vCuts = new ArrayList<Integer>(vCuts1);
		vCuts.add(grid[0].length);
		ArrayList<Integer> hCuts = new ArrayList<Integer>(hCuts1);
		hCuts.add(grid.length);
		
		int vPrev = 0, min=Integer.MAX_VALUE;
		for(int i=0; i<vCuts.size(); i++){
			int v = vCuts.get(i);
			int hPrev=0;
			for(int j=0; j<hCuts.size(); j++){				
				int h = hCuts.get(j);
				
				int thisSum=getSum(grid, hPrev,h-1, vPrev, v-1);
				System.out.println("This sum:" + thisSum);			
				min = Math.min(min, thisSum);
				hPrev = h;
			}
            System.out.println("---------------------");
			vPrev = v;
		}
		
		bestMinPiece = Math.max(min, bestMinPiece);
		System.out.println("Best min piece:" + bestMinPiece);
		System.out.println();
		System.out.println("---------------------");		
	}
	private static int getSum(int[][] grid, int rowStart, Integer rowEnd, int colStart, Integer colEnd) {
		System.out.println("Colstart:" + colStart + " Colend:" + colEnd);
		System.out.println("Rowstart:" + rowStart + " Rowend:"+rowEnd);
		
		if(rowEnd<rowStart)
		{
			System.out.println("That sucks!");
		}
		if(colEnd<colStart)
		{
			System.out.println("That sucks!");
		}
		int sum = 0;
		for(int i=rowStart; i<=rowEnd; i++){
			for(int j=colStart; j<=colEnd; j++){
				sum+=grid[i][j];
			}
		}
		return sum;		
	}

}

File: TopCoder/SRM/636/ChocolateDividingEasy/solution.py
import sys
import math

bestMinPiece=-1
def main():
	#2D-array named grid
	grid = [[9, 7, 6, 8],[6,7,6,7],[5,3,1,3]]
	
	numVertCuts = 2
	numHorzCuts = 2
	
	solve(grid,numVertCuts,numHorzCuts)
	print "Ans: " + str(bestMinPiece)
	
def solve(grid,numVertCuts,numHorzCuts):
	vCutsList = getCuts(len(grid),numVertCuts)
	hCutsList = getCuts(len(grid[0]),numHorzCuts)
	
	for vCuts in vCutsList:
		for hCuts in hCutsList:
			calcBestMinPiece(grid,vCuts,hCuts)
	
	return bestMinPiece

def getCuts(length,numCuts):
	set = []
	for j in range(1,length):
		set.append(j)
	
	results = []
	n = int(math.pow(2,length-1))
	for i in range(0,n):
		k=i
		index=0
		combo = []
		while (k):
			combo.append(set[index])
			k = k>>1
			index = index+1
		
		if len(combo) == numCuts:
			results.append(combo)
	
	return results

def calcBestMinPiece(grid,hCuts1,vCuts1):
	print "Vertical cuts:"
	for v in vCuts1:
		print str(v) + " "
	
	print ""
	
	print "Horizontal cuts:"
	for h in hCuts1:
		print str(h) + " "
	
	print ""
	print ""
	
	vCuts = []
	vCuts.extend(vCuts1);
	vCuts.append(len(grid[0]))
	hCuts = []
	hCuts.extend(hCuts1);
	hCuts.append(len(grid))
	
	vPrev=0
	min1 = sys.maxint
	for i in range(len(vCuts)):
		v = vCuts[i]
		hPrev = 0
		for j in range(len(hCuts)):
			h = hCuts[j]
			thisSum = getSum(grid,hPrev,h,vPrev,v)
			print "This sum:" + str(thisSum)
			min1 = min(min1,thisSum)			
			hPrev = h
		print "---------------------------"
		vPrev = v
	global bestMinPiece    
	bestMinPiece = max(min1, bestMinPiece)
	print "Calculated min: " + str(min1)
	print "Best min piece:" + str(bestMinPiece)
	print "----------------------------"

def getSum(grid,rowStart,rowEnd,colStart,colEnd):
	print "Colstart:" + str(colStart) + " Colend:" + str(colEnd)
	print "Rowstart:" + str(rowStart) + " Rowend:" + str(rowEnd)
	
	sum = 0;
	for i in range(rowStart,rowEnd):
		for j in range(colStart,colEnd):
			sum += grid[i][j]
	
	return sum
